### ISSUE TITLE
Planning for a sparse representation for the u_nk matrix

### ISSUE THREAD

**Issue by mrshirts** [choderalab/pymbar#285]
For local estimators (exact theory implemented still to be decided on), we may have u_kn matrices that have zero entries.  We need ways to store this information. The current storage process could be used with minor modifications (for example, some auxiliary arrays that describe which state transition energies are calculated), but would result in a large overuse if space.  For example, if we only calculate i+/-1 energies, and there are 40 states, approximately 37/40 = 93% of the space would be unused.  

One important questions we have to address.

1. Will the neighborhoods around each state change for different samples?   If, during the course of the analysis, we ONLY calculated i+/-1, and never calculated i+/-2, then we could make a number of assumptions that could significantly reduce the amount of information that needs to be stored.  It seems unlikely to come up with an algorithm that switched partway through, since you could imagine But if we wanted to combine data from one simulation performed with i+/-1 and another with i+/-2, then the neighborhoods would change between samples.

Some possibilities:

Currently we have an NxK array u_kn, that has for each sample, the energy of that sample in the  kth state, and a Kx1 array of the number of samples from each state.

1. Built-in sparse arrays.  This seems likely to cause lots of problems.  From https://docs.scipy.org/doc/scipy/reference/sparse.html: "Despite their similarity to NumPy arrays, it is strongly discouraged to use NumPy functions directly on these matrices because NumPy may not properly convert them for computations".  This makes me worried.

2. The most general thing to do would be to take the NxK array, and add a separate NxK boolean indicator if that state was used or not.  This would avoid needing to recognize a zero as not used or not.  But takes 2x as much space.   So doesn't help that much.  *However, I think that for initial testing of local methods, this is exactly what we should start with, since we don't really know how the different local methods will likely use the data, so we don't know what the cost of different representations will be*

3. We could have a N*(up to K) x 1 array of floats, giving the energies (energy array), along with a N*(up to K) x 1 array of ints (index array), which specifies at which state each energy was evaluated.  We would replace the Kx1 array (n_k) with a Nx1 matrix which has the state that each sample came from.  We can't use the N_k array, because we don't know how many samples for each state there are.  We can of course create one from the Nx1 matrix.   So for every energy, we know what state it came from, and what state it is the energy from.

For a simulation with 3 points, from each of the first states, evaluated at all of 5 states, the u_nk energy array would just be [energy of sample (0,0) in 5 states, energy of sample (0,1) in 5 states) . . . energies of sample (2,1) in 5 states] where sample (i,j) indicates the ith sample from the jth state.  

the index array would be [1,2,3,4,5] repeated 6 times. 

The state array would be [0,. . . [15 times] . . . 0, 1 . . . [15 times . . . ].  

Now let's say that each sample is just from i+/-1.  The energy array would have (where E[i] is the energy in state i) would be length 15 and have.
[E[0],E[1],E[0],E[1],E[0],E[1],E[0],E[1],E[2], E[0],E[1],E[2], E[0],E[1],E[2]]
The index array would be  [0,1,0,1,0,1,0,1,2,0,1,2,0,1,2] (also length 15)

And the state array would be: [0,0,0,0,0,0,1,1,1,1,1,1,1,1,1]

One issue with this representation is we don't actually know how many energies are associated with each sample, and slicing might be hard/time consuming.

4. We could also imagine, in addition to what is described in 3, a Nx(up to K) sample array which states what sample each energy term is from (in the example [0,0,1,1,2,2,3,3,3,4,4,4,5,5,5,6,6,6]).

5. The information in this sample array could also be stored as a Nx1 array of the number of energies associated with each sample, (in the example, [2,2,2,3,3,3]), in case that is needed anywhere)

One thing that I am seeing as we create this: we REALLY need to figure out how to make a representation that is not so hard for the user to understand.  It's been hard enough for people to figure out how to create the data for the u_kln matrix, and most of these sparse versions are much more complicated.  PERHAPS they store it in sparse matrix on data entry, and then we convert the sparse matrix an in internal representation like those described above to actually operate on it?

I will keep thinking about it.

**Comment by jchodera** [comment]
> The most general thing to do would be to take the NxK array, and add a separate NxK boolean indicator if that state was used or not. This would avoid needing to recognize a zero as not used or not. But takes 2x as much space. So doesn't help that much. However, I think that for initial testing of local methods, this is exactly what we should start with, since we don't really know how the different local methods will likely use the data, so we don't know what the cost of different representations will be

Sure, let's do this for now. This is how we represent things on disk for YANK right now anyway.

**Comment by mrshirts** [comment]
> To clarify: The initial implementation has already been decided on:

Yes. I'm just pointing out but we don't know how well it will work, whether it's possible to have a asymptotic variance estimator, or how to best implement it numerically within the code. Until we play around with things, we won't be able to really say what the best way to represent the data will be.

**Comment by mrshirts** [comment]
> Sure, let's do this for now. This is how we represent things on disk for YANK right now anyway.

Sounds like a good plan to start with to me!
